File: getstats.py
#!/usr/bin/env python

import re
import sys

class TextStat(object):
    def __init__(self):
        self.string_to_count = {}

    def update(self, c):
        self.string_to_count[c] = self.string_to_count.get(c, 0) + 1

    def summary(self):
        count = sum(self.string_to_count.values())
        null_count = self.string_to_count.get('', 0)
        strs = [s for s in self.string_to_count.keys() if len(s) != 0]
        lengths = [len(s) for s in strs]
        lengths.sort() 
        least_len = lengths[0]
        most_len = lengths[-1]
        shortest_str = sorted([s for s in strs if len(s) == least_len])[0]
        longest_str = sorted([s for s in strs if len(s) == most_len])[0]
        num_shortest = self.string_to_count[shortest_str]
        num_longest = self.string_to_count[longest_str]
        avg_len = sum(self.string_to_count[s] * len(s) for s in strs) / float(count - null_count)

        yield ' count:                  %s' % count
        yield ' null count:             %s' % null_count
        yield ' count(shortest str):    %s %s' % (num_shortest, shortest_str)
        yield ' count(longest str):     %s %s' % (num_longest, longest_str)
        yield ' average length:         %.2f' % avg_len


class NumberStat(object):
    def __init__(self):
        self.null_count = 0
        self.count = 0
        self.total = 0.0
        self.least = float('inf')
        self.most = float('-inf')

    def update(self, c):
        self.count += 1
        if c == '':
            self.null_count += 1
        else:
            num = float(c)
            self.total += num
            self.least = min(self.least, num)
            self.most = max(self.most, num)
    
    def summary(self):
        avg = self.total / float(self.count - self.null_count)
        yield ' count:      %s' % self.count
        yield ' null_count: %s' % self.null_count
        yield ' min:        %.3f' % self.least
        yield ' max:        %.3f' % self.most
        yield ' total:      %.3f' % self.total
        yield ' avg:        %.3f' % avg


field_type_rx = re.compile(r'\(([^)]*)\)')
header_clean_rx = re.compile(r'"|\(.*?\)')

header_line = sys.stdin.readline()
header_cleaned = header_clean_rx.sub('', header_line)
field_types = field_type_rx.findall(header_line)
field_names = [s.strip() for s in header_cleaned.split(',')]

def make_stat_for_type(t):
    if t == 'number':
        return NumberStat()
    else:
        return TextStat()

stats = [make_stat_for_type(t) for t in field_types]

for line in sys.stdin:
    line = line.strip()
    row = line.split(',')
    for c, stat in zip(row, stats):
        stat.update(c) 

for name, s in zip(field_names, stats):
    print '%s:' % name
    print '\n'.join(s.summary())
    print

File: getstats.go
package main

import "bufio"
import "fmt"
import "log"
import "math"
import "os"
import "regexp"
import "sort"
import "strconv"
import "strings"

type Stat interface {
	update(cell string, lineNum int)
	summarize()
}

type TextStat struct {
	stringToCount map[string]int
}

func makeTextStat() *TextStat {
	return &TextStat{stringToCount: make(map[string]int)}
}

func (ts *TextStat) update(cell string, lineNum int) {
	_, ok := ts.stringToCount[cell]
	if ok {
		ts.stringToCount[cell]++
	} else {
		ts.stringToCount[cell] = 1
	}
}

func (ts *TextStat) summarize() {
	strs := make([]string, 0)
	lengths := make([]int, 0)
	count := 0
	nullCount := 0
	for s, c := range ts.stringToCount {
		count += c
		if s == "" {
			nullCount += c
		} else {
			strs = append(strs, s)
			lengths = append(lengths, len(s))
		}
	}
	sort.Ints(lengths)
	leastLen := lengths[0]
	mostLen := lengths[len(lengths)-1]

	shortestStrs := make([]string, 0)
	longestStrs := make([]string, 0)
	for _, s := range strs {
		if len(s) == leastLen {
			shortestStrs = append(shortestStrs, s)
		}
		if len(s) == mostLen {
			longestStrs = append(longestStrs, s)
		}
	}

	// Break ties alphabetically
	sort.Strings(shortestStrs)
	sort.Strings(longestStrs)
	shortestStr := shortestStrs[0]
	longestStr := longestStrs[0]

	numShortest := ts.stringToCount[shortestStr]
	numLongest := ts.stringToCount[longestStr]

	lenTotal := 0
	for s, c := range ts.stringToCount {
		lenTotal += c * len(s)
	}
	avgLen := float32(lenTotal) / float32(count-nullCount)

	fmt.Printf(" count:                  %d\n", count)
	fmt.Printf(" null count:             %d\n", nullCount)
	fmt.Printf(" count(shortest str):    %d %s\n", numShortest, shortestStr)
	fmt.Printf(" count(longest str):     %d %s\n", numLongest, longestStr)
	fmt.Printf(" average length:         %.2f\n", avgLen)
}

type NumberStat struct {
	nullCount int
	count     int
	total     float64
	least     float64
	most      float64
}

func makeNumberStat() *NumberStat {
	return &NumberStat{least: math.Inf(1), most: math.Inf(-1)}
}

func (self *NumberStat) update(cell string, lineNum int) {
	self.count++
	if cell == "" {
		self.nullCount++
	} else {
		num, err := strconv.ParseFloat(cell, 64)
		if err != nil {
			log.Fatalf("Failed to parse supposed float '%s' at line %d\n", cell, lineNum)
		}
		self.total += num
		self.least = math.Min(num, self.least)
		self.most = math.Max(num, self.most)
	}
}

func (self *NumberStat) summarize() {
	avg := self.total / float64(self.count-self.nullCount)
	fmt.Printf(" count:      %d\n", self.count)
	fmt.Printf(" null_count: %d\n", self.nullCount)
	fmt.Printf(" min:        %.3f\n", self.least)
	fmt.Printf(" max:        %.3f\n", self.most)
	fmt.Printf(" total:      %.3f\n", self.total)
	fmt.Printf(" avg:        %.3f\n", avg)
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	// Read the header line, get the field names and types
	fieldTypeRx := regexp.MustCompile(`\(([^)]*)\)`)
	headerCleanRx := regexp.MustCompile(`"|\(.*?\)`)
	headerLine, err := reader.ReadString('\n')
	if err != nil {
		log.Fatalln("Failed to read header line.", err)
	}
	fieldTypes := fieldTypeRx.FindAllString(headerLine, -1)
	headerCleaned := headerCleanRx.ReplaceAllString(headerLine, "")
	fieldNames := strings.Split(headerCleaned, ",")
	for i, name := range fieldNames {
		fieldNames[i] = strings.TrimSpace(name)
	}

	// Init the stats
	stats := make([]Stat, len(fieldNames))
	for i, _ := range fieldNames {
		if fieldTypes[i] == "(number)" {
			stats[i] = makeNumberStat()
		} else {
			stats[i] = makeTextStat()
		}
	}

	// Read lines from stdin, updating the stats as we go
	for lineNum := 2; ; lineNum++ {
		line, err := reader.ReadString('\n')
		if err != nil {
			break
		}
		fields := strings.Split(line, ",")
		for i, field := range fields {
			field = strings.TrimSpace(field)
			stats[i].update(field, lineNum)
		}
	}

	// Print out the summary statistics
	for i, fieldName := range fieldNames {
		fmt.Printf("%s:\n", fieldName)
		stats[i].summarize()
		fmt.Println()
	}
}
